tokio-timer/src/throttle.rs

```diff
@@ -0,0 +1,68 @@
+//! Slow down a stream by enforcing a delay between items.
+
+use {Delay, Error};
+
+use futures::{Async, Future, Poll, Stream};
+use futures::future::Either;
+
+use std::time::{Duration, Instant};
+
+/// Slow down a stream by enforcing a delay between items.
+#[derive(Debug)]
+pub struct Throttle<T> {
+    delay: Option<Delay>,
+    duration: Duration,
+    stream: Option<T>,
+}
+
+impl<T> Throttle<T> {
+    /// Slow down a stream by enforcing a delay between items.
+    pub fn new(stream: T, duration: Duration) -> Self {
+        Self {
+            delay: None,
+            duration: duration,
+            stream: Some(stream),
+        }
+    }
+}
+
+impl<T: Stream> Stream for Throttle<T> {
+    type Item = T::Item;
+    type Error = Either<T::Error, Error>;
```

Comment:
For forwards compat, I would define `struct ThrottleError<T>(Either<T, Error>)`

Reply:
But `Either` would still be public, right?

Reply:
I would keep it private.

I wouldn't worry about providing access to the inner error until someone asks for it :)

Reply:
This is the strategy i use for errors that contain multiple variants: https://github.com/carllerche/h2/issues/310

Reply:
hmm, can't really use this in pattern matching, can you?